core/include/drivers/rstctrl.h

```diff
@@ -0,0 +1,217 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, Linaro Limited
+ *
+ * Self contained reset controller framework in rstctrl.h
+ * A reset controller exposes reset level controls to a
+ * platform.
+ */
+
+#ifndef __DRIVERS_RSTCTRL_H
+#define __DRIVERS_RSTCTRL_H
+
+#include <kernel/dt.h>
+#include <kernel/dt_driver.h>
+#include <stdint.h>
+#include <tee_api_types.h>
+
+struct rstctrl;
+
+struct rstctrl_ops {
+	/*
+	 * Operators on reset control(s) exposed by a reset controller
+	 *
+	 * assert_level: Assert reset level on control with a timeout hint
+	 * deassert_level: Deassert reset level on control with a timeout hint
+	 * get_name: Get a string name for the controller, or NULL is none
+	 *
+	 * @rstctrl: Reset control reference
+	 * @id: Identifier for the reset level control in the reset controller
+	 * @to_ms: Timeout in microsencond or RSTCTRL_NO_TIMEOUT, may be ignored
+	 * by reset controller.
+	 * Return a TEE_Result compliant code
+	 */
+	TEE_Result (*assert_level)(struct rstctrl *rstctrl, unsigned int to_us);
+	TEE_Result (*deassert_level)(struct rstctrl *rstctrl,
+				     unsigned int to_us);
+	const char *(*get_name)(struct rstctrl *rstctrl);
+};
+
+/*
+ * struct rstctrl - Instance of a control exposed by a reset controller
+ * @ops: Operators of the reset controller
+ * @priv: Reset control private data (unsigned value or memory pointer)
+ * @exclusive: Set when a consumer has exclusive control on the reset level
+ */
+struct rstctrl {
+	const struct rstctrl_ops *ops;
+	uintptr_t priv;
+	bool exclusive;
```

Comment:
How will this field be used? Is it needed for all instances?

Reply:
In case a reset controller is shared by several devices, one may request exclusive access to the controller when using the related resource and release exclusive access once resource is no more used. Likely needed only for specific shared reset controllers. I've added here to have a generic structure.

Reply:
> How will this field be used? 

Sorry, I didn't answer that question. It is used through `rstctrl_{get|put}_exclusive()`.

Reply:
Thanks